PYME/cluster/rules.py

```diff
@@ -0,0 +1,469 @@
+
+import time
+import threading
+from PYME.IO import DataSources, clusterIO, unifiedIO, clusterResults
+import json
+import socket
+import random
+from PYME.misc import hybrid_ns
+from PYME.misc.computerName import GetComputerName
+import dispatch
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+def _get_ruleserver_uri(n_retries=2):
+    """Discover the distributors using zeroconf and choose one"""
+    ns = hybrid_ns.getNS('_pyme-taskdist')
+
+    servers = {}
+
+    def _search():
+        for name, info in ns.get_advertised_services():
+            if name.startswith('PYMERuleServer'):
+                print(info, info.address)
+                servers[name] = 'http://%s:%d' % (socket.inet_ntoa(info.address), info.port)
+
+    _search()
+    while not servers and (n_retries > 0):
+        logging.info('could not find a rule server, waiting 5s and trying again')
+        time.sleep(5)
+        n_retries -= 1
+        _search()
+
+    # if we only have one option, go with it
+    server_uris = list(servers.values())
+    if len(server_uris) == 1:
+        return server_uris[0]
+    # otherwise, prefer a local server if available
+    try:
+        return servers[GetComputerName()]
+    except KeyError:
+        logging.info('no local rule server; found %d, choosing one at random' % len(server_uris))
+        return random.choice(server_uris)
+
+
+class Rule(object):
+    def __init__(self):
+        self._template = {'id': '{{ruleID}}~{{taskID}}'}
+        self._on_completion = None
+        self.ruleserver_uri = None
+        self._rule_id = None
+        self._max_tasks = 1
+
+    @property
+    def template(self):
+        return self._template
+
+    @property
+    def rule(self):
+        rule = {'template': json.dumps(self.template)}  # todo - change rulenodeserver so we don't have to dumps template first
```

Comment:
The template should be a template string. This is not going to change on the ruleserver.

Reply:
Works for me, let's take out the comment then